EKF/control.cpp

```diff
@@ -744,30 +747,69 @@ void Ekf::controlGpsYawFusion()
 		return;
 	}
 
-	if (ISFINITE(_gps_sample_delayed.yaw)) {
+	const bool is_new_data_available = ISFINITE(_gps_sample_delayed.yaw);
+
+	if (is_new_data_available) {
+
+		const bool continuing_conditions_passing = !gps_checks_failing;
+
+		const bool is_gps_yaw_data_intermittent = !isRecent(_time_last_gps_yaw_data, 2 * GPS_MAX_INTERVAL);
+
+		const bool starting_conditions_passing = continuing_conditions_passing
+		                                         && _control_status.flags.tilt_align
+		                                         && gps_checks_passing
+							 && !is_gps_yaw_data_intermittent
+		                                         && !_gps_hgt_intermittent;
+
+		_time_last_gps_yaw_data = _time_last_imu;
 
 		if (_control_status.flags.gps_yaw) {
-			fuseGpsYaw();
 
-		} else {
-			// Try to activate GPS yaw fusion
-			if (_control_status.flags.tilt_align
-			    && !_gps_hgt_intermittent) {
+			if (continuing_conditions_passing) {
 
+				fuseGpsYaw();
+
+				const bool is_fusion_failing = isTimedOut(_time_last_gps_yaw_fuse, _params.reset_timeout_max);
+
+				if (is_fusion_failing) {
+					if (_nb_gps_yaw_reset_available > 0) {
+						// Data seems good, attempt a reset
+						resetYawToGps();
+						_nb_gps_yaw_reset_available--;
+
+					} else if (starting_conditions_passing) {
+						// Data seems good, but previous reset did not fix the issue
+						// something else must be wrong, declare the sensor faulty and stop the fusion
+						_is_gps_yaw_faulty = true;
+						stopGpsYawFusion();
+
+					} else {
+						// A reset did not fix the issue but all the starting checks are not passing
+						// This could be a temporary issue, stop the fusion without declaring the sensor faulty
+						stopGpsYawFusion();
+					}
+					// TODO: should we give a new reset credit when the fusion does not fail for some time?
+				}
+
+			} else {
+				// Stop GPS yaw fusion but do not declare it faulty
+				stopGpsYawFusion();
+			}
+
+		} else {
+		        if (starting_conditions_passing) {
+				// Try to activate GPS yaw fusion
 				if (resetYawToGps()) {
 					_control_status.flags.yaw_align = true;
+					_nb_gps_yaw_reset_available = 1;
 					startGpsYawFusion();
 				}
 			}
 		}
-	}
 
-	// Check if the data is constantly fused by the estimator,
-	// if not, declare the sensor faulty and stop the fusion
-	// By doing this, another source of yaw aiding is allowed to start
-	if (_control_status.flags.gps_yaw
-	    && isTimedOut(_time_last_gps_yaw_fuse, (uint64_t)5e6)) {
-		_is_gps_yaw_faulty = true;
+	} else if (_control_status.flags.gps_yaw
+	           && isTimedOut(_time_last_gps_yaw_data, _params.reset_timeout_max)) {
+		// No yaw data in the message anymore. Stop until it comes back.
 		stopGpsYawFusion();
```

Comment:
This is an interesting case and one that can persist for some time when on ground. I have seen logs where the GNSS yaw is acquired, glitches, then stops for up to a minute until the RTK functionality is restored. The code looks like it should fuse _last_static_yaw if _control_status.flags.vehicle_at_rest is true to prevent uncontrolled yaw drift and enable yaw gyro bias learning in this period. This is something we should check when looking at a test log.

Reply:
If the mag fusion is set to NONE, then it will enter this case and fuse the last static yaw if on ground:
https://github.com/PX4/PX4-ECL/blob/29243ac5cbb5d27ac71744e88afcd786df6f748d/EKF/mag_control.cpp#L49-L59

Otherwise, it will simply start the desired mag fusion mode

Reply:
I added a unit test for this specific case here: https://github.com/PX4/PX4-ECL/pull/1006/commits/400b66d76bee133d50acf0343b7af489a0c1b74a

Reply:
However, if there is no mag data and the `MAG_FUSE_TYPE` is not set to `NONE` (i.e.: auto, 3d or heading), the yaw variance is not constrained. This is maybe something we need to address as well in a following PR